cognite_toolkit/_cdf_tk/utils/fileio/_compression.py

```diff
@@ -0,0 +1,78 @@
+
+
+class Compression(ABC):
+    encoding = "utf-8"
+    newline = "\n"
+    name: ClassVar[str]
+    file_suffix: ClassVar[str]
+
+    def __init__(self, filepath: Path) -> None:
+        self.filepath = filepath
+
+    @abstractmethod
+    def open(self, mode: Literal["r", "w"]) -> TextIOWrapper:
+        """Open the compressed file and return a file-like object."""
+        raise NotImplementedError("This method should be implemented in subclasses.")
+
+    @classmethod
+    def from_filepath(cls, filepath: Path) -> "Compression":
+        return COMPRESSION_BY_SUFFIX.get(filepath.suffix, NoneCompression)(filepath=filepath)
+
+    @classmethod
+    def from_name(cls, compression: str) -> "type[Compression]":
+        if compression in COMPRESSION_BY_NAME:
+            return COMPRESSION_BY_NAME[compression]
+        raise ToolkitValueError(
+            f"Unknown compression type: {compression}. Available types: {humanize_collection(COMPRESSION_BY_NAME.keys())}."
+        )
+
+
+class NoneCompression(Compression):
+    name = "none"
+    file_suffix = ""
+
+    def open(self, mode: Literal["r", "w"]) -> TextIOWrapper:
+        """Open the file without compression."""
+        return self.filepath.open(mode=mode, encoding=self.encoding, newline=self.newline)
+
+
+class GzipCompression(Compression):
+    name = "gzip"
+    file_suffix = ".gz"
+
+    def open(self, mode: Literal["r", "w"]) -> TextIOWrapper:
+        """Open the gzip compressed file."""
+        # MyPy (or gzip) fails to recognize that gzip.open returns a TextIOWrapper
+        return gzip.open(self.filepath, mode=f"{mode}t", encoding=self.encoding, newline=self.newline)  # type: ignore[return-value]
+
+
+COMPRESSION_BY_SUFFIX: Mapping[str, type[Compression]] = {}
+COMPRESSION_BY_NAME: Mapping[str, type[Compression]] = {}
+for subclass in get_concrete_subclasses(Compression):  # type: ignore[type-abstract]
+    if subclass.name in COMPRESSION_BY_NAME:
+        raise TypeError(
+            f"Duplicate compression name {subclass.name!r} found for classes "
+            f"{COMPRESSION_BY_NAME[subclass.name].__name__!r} and {subclass.__name__!r}."
+        )
+    # We use `type: ignore[index]` to suppress MyPy errors about the type of the dictionary.
+    # This is because we want to expose COMPRESSION_BY_NAME and COMPRESSION_BY_SUFFIX as
+    # `Mapping[str, type[Compression]]`, while we know they are actually dictionaries.
+    COMPRESSION_BY_NAME[subclass.name] = subclass  # type: ignore[index]
+    if not subclass.file_suffix:
+        continue
+    if subclass.file_suffix in COMPRESSION_BY_SUFFIX:
+        raise TypeError(
+            f"Duplicate compression file suffix {subclass.file_suffix!r} found for classes "
+            f"{COMPRESSION_BY_SUFFIX[subclass.file_suffix].__name__!r} and {subclass.__name__!r}."
+        )
+    COMPRESSION_BY_SUFFIX[subclass.file_suffix] = subclass  # type: ignore[index]
```

Comment:
Do you expect there to be a lot of these Compression classes?
It seems a bit over-engineered to use runtime reflection to find subclasses of `Compression` when the list of possible compression methods is probably small?

`__subclasses__` will only find subclasses that have been loaded into memory, so there is still a chance that some classes will be missed if someone refactors and moves some compression classes to another file.

It seems more robust to just use a list
```suggestion
_all_compression_types: list[type[Compression]] = [NoneCompression, GzipCompression]
COMPRESSION_BY_SUFFIX: Mapping[str, type[Compression]] = {x.file_suffix: x for x in _all_compression_types}
COMPRESSION_BY_NAME: Mapping[str, type[Compression]] = {x.name: x for x in _all_compression_types}
assert len(_all_compression_types) == len(COMPRESSION_BY_SUFFIX), "Duplicated compression suffix"
assert len(_all_compression_types) == len(COMPRESSION_BY_NAME), "Duplicated compression name"
```

What do you think? 🤔

Reply:
I do expect a couple more. I typically will have it all in the same file, or just move the subclass call. However it doesn't hurt with a test

Reply:
But is it more readable and robust to hard code a list of compression classes?

Reply:
I am not suggesting an implementation without classes, but one without reflection. Like the example above.

Reply:
I like reflection, it is a pattern  we use a lot in Toolkit. I like that we automatically register all classes. I have used the hardcoded variant before and experienced that it gets outdated.